util/selection/selection.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package selection
+
+import (
+	"math/rand"
+	"sort"
+)
+
+// Interface is alias of sort.Interface
+type Interface = sort.Interface
+
+// Select performs introselect algorithm on data and return index of the k-th smallest value.
+func Select(data Interface, k int) int {
+	if data.Len() > 0 {
+		return introselect(data, 0, data.Len()-1, k-1, 0)
+	}
+	return -1
+}
+
+// introselect will perform quickselect at beginning, and switch to linear-time algorithm if it recurses too much times.
+// Source paper: http://www.cs.rpi.edu/~musser/gp/introsort.ps
+func introselect(data Interface, left, right, k int, depth int) int {
+	if left == right {
+		return left
+	}
+	if depth > 6 {
+		// Use median of medians algorithm(linear-time selection) when recurses too much times.
+		return medianOfMedians(data, left, right, k)
+	}
+	// TODO: use a better pivot function
+	pivotIndex := randomPivot(data, left, right)
+	pivotIndex = partition(data, left, right, pivotIndex)
```

Comment:
Why not binary divide here?

Reply:
Random pivot can avoid median of 3 killer, and performs as well as binary pivot in most cases.

Morever, k is stable during a select prodedure, in which case binary pivot may cause more partition operations.

We can optimize this later since it's good enough for now that it can always be faster than do a full sort with `sort.Sort` of golang.